libsyclinterface/source/dpctl_sycl_kernel_bundle_interface.cpp

```diff
@@ -0,0 +1,734 @@
+
+#include "dpctl_sycl_kernel_bundle_interface.h"
+#include "Config/dpctl_config.h"
+#include "Support/CBindingWrapping.h"
+#include "dpctl_dynamic_lib_helper.h"
+#include "dpctl_error_handlers.h"
+#include <CL/cl.h>     /* OpenCL headers     */
+#include <CL/sycl.hpp> /* Sycl headers       */
+#include <CL/sycl/backend/opencl.hpp>
+#include <sstream>
+
+#ifdef DPCTL_ENABLE_L0_PROGRAM_CREATION
+// Note: include ze_api.h before level_zero.hpp. Make sure clang-format does
+// not reorder the includes.
+// clang-format off
+#include "ze_api.h" /* Level Zero headers */
+#include "sycl/ext/oneapi/backend/level_zero.hpp"
+// clang-format on
+#endif
+
+using namespace cl::sycl;
+
+namespace
+{
+DEFINE_SIMPLE_CONVERSION_FUNCTIONS(context, DPCTLSyclContextRef)
+DEFINE_SIMPLE_CONVERSION_FUNCTIONS(device, DPCTLSyclDeviceRef)
+DEFINE_SIMPLE_CONVERSION_FUNCTIONS(kernel_bundle<bundle_state::executable>,
+                                   DPCTLSyclKernelBundleRef)
+DEFINE_SIMPLE_CONVERSION_FUNCTIONS(kernel, DPCTLSyclKernelRef)
+
+#ifdef __linux__
+static const char *clLoaderName = DPCTL_LIBCL_LOADER_FILENAME;
+static const int clLibLoadFlags = RTLD_NOLOAD | RTLD_NOW | RTLD_LOCAL;
+#elif defined(_WIN64)
+static const char *clLoaderName = "OpenCL.dll";
+static const int clLibLoadFlags = 0;
+#else
+#error "OpenCL program compilation is unavailable for this platform"
+#endif
+
+#define CodeStringSuffix(code)                                                 \
+    std::string(" (code=") + std::to_string(static_cast<int>(code)) + ")"
+
+#define EnumCaseString(code)                                                   \
+    case code:                                                                 \
+        return std::string(#code) + CodeStringSuffix(code)
+
+constexpr backend cl_be = backend::opencl;
+
+struct cl_loader
+{
+public:
+    static cl_loader &get()
+    {
+        static cl_loader _loader;
+        return _loader;
+    }
+
+    template <typename retTy> retTy getSymbol(const char *name)
+    {
+        if (!opened) {
+            error_handler("The OpenCL loader dynamic library could not "
+                          "be opened.",
+                          __FILE__, __func__, __LINE__);
+
+            return nullptr;
+        }
+        return clLib.getSymbol<retTy>(name);
+    }
+
+private:
+    dpctl::DynamicLibHelper clLib;
+    bool opened;
+    cl_loader() : clLib(clLoaderName, clLibLoadFlags), opened(clLib.opened()) {}
+};
+
+typedef cl_program (*clCreateProgramWithSourceFT)(cl_context,
+                                                  cl_uint,
+                                                  const char **,
+                                                  const size_t *,
+                                                  cl_int *);
+const char *clCreateProgramWithSource_Name = "clCreateProgramWithSource";
+clCreateProgramWithSourceFT get_clCreateProgramWithSource()
+{
+    static auto st_clCreateProgramWithSourceF =
```

Comment:
Maybe we should add a check here that the symbol was found?

Reply:
There is a check already in the `DynamicLibHelper`` method. Additionally there is a check after each call to `get_SymbNameF` function.